### DISCUSSION TITLE
Multibyte character string of POST data is garbled using "fastapi.Form".

### DISCUSSION THREAD

**Discussion by delhi09** [fastapi/fastapi#7071] [Questions] [answered]
I am Japanese.

During studying and development with the FastAPI, I found that the multibyte string of POST data (in my case Japanese) is garbled using "fastapi.Form".
https://fastapi.tiangolo.com/tutorial/request-forms/

### Example
Write the following Fast API code.

```Python
from fastapi import APIRouter, Form

router = APIRouter()

@router.post("/")
async def create_xxx(name: str = Form(..., max_length=255)):
    # When I execute the curl command described below,
    # I found the "name" parameter's value is garbled by putting a breakpoint.
```

Execute the curl command as shown below.

```Shell
$ curl -X POST --data-urlencoded "name=あいうえお" "http://localhost/example/"
```

```Shell
# The same result was obtained when the Content-Type was specified.
$ curl -X POST -H "Content-Type: application/x-www-form-urlencoded; charset=utf-8" --data-urlencode "name=あいうえお" "http://localhost/example/"
```

### Description
I found that the multibyte string of POST data (in my case Japanese) is garbled using "fastapi.Form".

When the code shown in the above "Example" is executed, the multibyte character string got on the API side is garbled.

I have posted the evidence below.
![evidence](https://user-images.githubusercontent.com/63476957/100513806-44e99d80-31b3-11eb-8e50-bda834ed48da.png)

At first, as a result of debugging, I thought that it was not a problem with the source code of FastAPI but a problem on the starlette side on which FastAPI depends, so I raised the following issue on the starlette forum.
[Non-latin-1 character string is garbled in POSTed data](https://discuss.encode.io/t/non-latin-1-character-string-is-garbled-in-posted-data/979)

For details, see the issue on the starlette forum, but starlette decodes the value of the parameter "Content-Type: application/x-www-form-urlencoded;" with "latin-1", ignoring the character code in the http request header, and I have confirmed that garbled characters will not occur if I locally modify the part that is decoded with "latin-1" to decode with "utf-8".

I would like to confirm the following in the FastAPI community.

1. Is my idea that it is a problem on the starlette side correct?
2. Is there a way to handle it on the FastAPI side?(by modifying settings or code)

※ I have confirmed that this issue does not occur when sending POST data in JSON format and receiving parameters using  "pydantic.BaseModel".
https://fastapi.tiangolo.com/tutorial/body/

### Environment

* OS: MacOS Catalina 10.15.7
* FastAPI version: 0.61.2
* starlette version:  0.13.6
* Python version: 3.7.3

**Comment by oryba** [comment, marked as answer]
> Is my idea that it is a problem on the starlette side correct?

I checked the sources too and found the same place where Starlette decodes field value as 'latin-1':
```python3
name = unquote_plus(field_name.decode("latin-1"))
value = unquote_plus(field_value.decode("latin-1"))
```

But there's one more thing worth mentioning. `unquote_plus(field_value.decode("latin-1"))` means that:
1. the **percent-encoded** result is decoded as `latin-1` (which is ok since percent-encoded string contains only valid `latin-1`)
```
>>> field_value
b'%E3%81%82%E3%81%84%E3%81%86%E3%81%88%E3%81%8A'
>>> field_value.decode('latin-1')
'%E3%81%82%E3%81%84%E3%81%86%E3%81%88%E3%81%8A'
```

So theoretically **there should be no problem with decoding utf-8 strings**.

I checked your example (slightly modified it by wrapping the route in an app), works fine with the test string you provided.

```python3
import uvicorn

from fastapi import APIRouter, Form, FastAPI

app = FastAPI()

router = APIRouter()

@router.post("/")
async def create_xxx(name: str = Form(..., max_length=255)):
    assert name == "あいうえお"

app.include_router(router, prefix='/example')

uvicorn.run(app, host='127.0.0.1', port=8000)
```

Test it:
```bash
curl -X POST -H "Content-Type: application/x-www-form-urlencoded;" --data-urlencode "name=あいうえ お" "http://127.0.0.1:8000/example/"
```

And the assertion is passed, so at least that works on my machine™:) Could you re-check or correct me if I missed something?

**Comment by delhi09** [comment]
Thank you for replying.

I re-checked and as you say, I confirmed that garbled characters do not occur when "--data-urlencode" is added to curl command.

If I do not add "--data-urlencode" to curl command,  grabled characters occur, so I've probably misunderstood that I added "--data-urlencode" and confirmed it.

I'm sorry for the careless mistake.

I understand that the code decoding with latin-1 is decoding the byte sequences into a URL-encoded string, which is correct code.

Then, "urllib.parse.unquote_plus" decode URL-encoded string with utf-8(default).
https://docs.python.org/3/library/urllib.parse.html#urllib.parse.unquote_plus

I close this issue.

**Comment by tiangolo** [comment]
Thanks for the help here @oryba ! 🍰 

And thanks for coming back to close the issue @delhi09 ☕ 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Reply by fanybook** [reply]
@tiangolo 

Why not decode it directly in the request.form()

https://github.com/fastapi/fastapi/discussions/13248
